File: al_qubaisi_modifications/al_qubaisi_modifications/report/employee_document_expiry/employee_document_expiry.py
# Copyright (c) 2026, Acube Innoivations Pvt Limited and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import add_days, getdate, nowdate


def execute(filters=None):
	filters = frappe._dict(filters or {})
	document_types = get_document_types()
	columns = get_columns(document_types)
	data = get_data(filters, document_types)
	return columns, data


def get_document_types():
	return frappe.get_all(
		"Employee Document Type",
		filters={"has_expiry": 1},
		pluck="name",
		order_by="name",
	)


def get_columns(document_types):
	columns = [
		{
			"fieldname": "employee",
			"label": _("Employee"),
			"fieldtype": "Link",
			"options": "Employee",
			"width": 120,
		},
		{
			"fieldname": "employee_name",
			"label": _("Employee Name"),
			"fieldtype": "Data",
			"width": 180,
		},
		{
			"fieldname": "company",
			"label": _("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"width": 140,
		},
		{
			"fieldname": "department",
			"label": _("Department"),
			"fieldtype": "Link",
			"options": "Department",
			"width": 140,
		},
		{
			"fieldname": "designation",
			"label": _("Designation"),
			"fieldtype": "Data",
			"width": 130,
		},
	]
	for document_type in document_types:
		columns.append(
			{
				"fieldname": frappe.scrub(document_type),
				"label": _("{0} Expiry").format(document_type),
				"fieldtype": "Date",
				"width": 130,
			}
		)
	return columns


def get_data(filters, document_types):
	conditions = ["ifnull(parent.employee_status, 'Active') = 'Active'"]
	values = {}
	if filters.get("company"):
		conditions.append("parent.company = %(company)s")
		values["company"] = filters.company
	if filters.get("department"):
		conditions.append("parent.department = %(department)s")
		values["department"] = filters.department

	rows = frappe.db.sql(
		"""
		select parent.employee, parent.employee_name, parent.company,
			parent.department, parent.designation,
			child.document_type, child.expiry_date
		from `tabEmployee Documents` parent
		left join `tabEmployee Documents Detail` child on child.parent = parent.name
		where {conditions}
		order by parent.employee_name
		""".format(conditions=" and ".join(conditions)),
		values,
		as_dict=True,
	)

	employees = {}
	for row in rows:
		record = employees.setdefault(
			row.employee,
			{
				"employee": row.employee,
				"employee_name": row.employee_name,
				"company": row.company,
				"department": row.department,
				"designation": row.designation,
			},
		)
		if row.document_type in document_types and row.expiry_date:
			fieldname = frappe.scrub(row.document_type)
			# keep the earliest expiry if the same document appears twice
			if not record.get(fieldname) or getdate(row.expiry_date) < getdate(record[fieldname]):
				record[fieldname] = row.expiry_date

	data = list(employees.values())

	if filters.get("expiring_within_days"):
		cutoff = add_days(getdate(nowdate()), int(filters.expiring_within_days))
		fieldnames = [frappe.scrub(dt) for dt in document_types]
		data = [
			record
			for record in data
			if any(record.get(f) and getdate(record[f]) <= cutoff for f in fieldnames)
		]

	return data

File: al_qubaisi_modifications/al_qubaisi_modifications/report/employee_document_expiry/employee_document_expiry.js
// Copyright (c) 2026, Acube Innoivations Pvt Limited and contributors
// For license information, please see license.txt

frappe.query_reports["Employee Document Expiry"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
		},
		{
			fieldname: "department",
			label: __("Department"),
			fieldtype: "Link",
			options: "Department",
		},
		{
			fieldname: "expiring_within_days",
			label: __("Expiring Within (Days)"),
			fieldtype: "Int",
			description: __("Only show employees with at least one document expiring within this many days (includes already expired)"),
		},
	],

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldtype === "Date" && data && data[column.fieldname]) {
			const expiry = frappe.datetime.str_to_obj(data[column.fieldname]);
			const today = frappe.datetime.str_to_obj(frappe.datetime.get_today());
			const soon = frappe.datetime.str_to_obj(
				frappe.datetime.add_days(frappe.datetime.get_today(), 30)
			);
			if (expiry < today) {
				value = `<span style="color: var(--red-600); font-weight: bold;">${value}</span>`;
			} else if (expiry <= soon) {
				value = `<span style="color: var(--orange-600); font-weight: bold;">${value}</span>`;
			}
		}
		return value;
	},
};
